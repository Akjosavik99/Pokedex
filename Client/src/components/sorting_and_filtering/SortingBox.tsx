import * as React from "react";
import Button from "@mui/material/Button";
import Menu from "@mui/material/Menu";
import MenuItem from "@mui/material/MenuItem";
import Fade from "@mui/material/Fade";
import { Box } from "@mui/material";
import styled from "@mui/material/styles/styled";
import sortings from "../../assets/Sortings";
import ArrowDropDownCircleOutlinedIcon from "@mui/icons-material/ArrowDropDownCircleOutlined";

type SortingBoxType = {
  setCurrentSorting: React.Dispatch<React.SetStateAction<string>>;
};

export default function SortingBox({ setCurrentSorting }: SortingBoxType) {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);

  const open = Boolean(anchorEl);

  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };
  const StyledMenuItem = styled(MenuItem)(() => ({
    "&.Mui-focusVisible": {
      backgroundColor: "lightblue",
    },
  }));

  const handleFocus = (text: string) => {
    const speechSynthesis = window.speechSynthesis;
    if (!speechSynthesis) return;
    speechSynthesis.cancel();
    const utterance = new SpeechSynthesisUtterance(text);
    utterance.volume = 0.5;
    speechSynthesis.speak(utterance);
  };

  return (
    <Box>
      <Button
        id="fade-button"
        aria-controls={open ? "fade-menu" : undefined}
        aria-haspopup="true"
        aria-expanded={open ? "true" : undefined}
        onClick={handleClick}
        data-testid="sort-list-button"
        sx={{
          backgroundColor: "primary.main",
          color: "white",
          padding: "10px",
          "&:hover": {
            backgroundColor: "primary.light",
          },
        }}
      >
        Choose Sorting{" "}
        <ArrowDropDownCircleOutlinedIcon
          sx={{ marginTop: "-5px", marginLeft: "5px" }}
        />
      </Button>
      <Menu
        id="fade-menu"
        MenuListProps={{
          "aria-labelledby": "fade-button",
        }}
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        TransitionComponent={Fade}
      >
        {sortings.map((sorting) => (
          <StyledMenuItem
            onFocus={() => handleFocus(sorting[0])}
            key={sorting[1]}
            value={sorting[1]}
            onKeyDown={(event) => {
              if (event.key === "Enter") {
                handleClose();
                setCurrentSorting(sorting[1]);
              }
            }}
            data-testid={sorting[0]}
            onClick={() => {
              handleClose();
              setCurrentSorting(sorting[1]);
            }}
            // style={{
            //   backgroundColor:
            //     sorting[1] === currentSorting ? "lightblue" : "white",
            // }}
          >
            {sorting[0]}
          </StyledMenuItem>
        ))}
      </Menu>
    </Box>
  );
}
