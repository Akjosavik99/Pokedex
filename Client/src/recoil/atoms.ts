import { atom } from "recoil";

/**
 * Global function that removes a filter from the filterBy array in session storage
 * @param filter
 * @returns filterBy array after removing the given filter
 */
export function removeFromFilter(filter: string) {
  const filterBy: string[] = JSON.parse(sessionStorage.getItem("filterBy")!);
  const index = filterBy.indexOf(filter);
  if (index > -1) {
    filterBy.splice(index, 1);
  }
  sessionStorage.setItem("filterBy", JSON.stringify(filterBy));
  return filterBy;
}

/**
 * Global state: MyTeam loaded from local storage
 */
export const recoilMyTeam = atom({
  key: "myTeamAtom",
  default: getTeam() as string[],
});

/**
 * Global state: MyTeam loaded from local storage
 */
function getTeam() {
  const team: string[] = JSON.parse(localStorage.getItem("team")!);
  if (!team) {
    localStorage.setItem("team", JSON.stringify([]));
    return [];
  }
  return team;
}

/**
 * Global state: Filter by
 */
export const recoilFilterBy = atom<string[]>({
  key: "filterByAtom",
  default: [],
});

/**
 * Global state: Sort by
 */
export const recoilSortBy = atom<string>({
  key: "sortByAtom",
  default: "_id,1",
});

/**
 * Global state: Search string
 */
export const recoilSearch = atom<string>({
  key: "searchAtom",
  default: "",
});

/**
 * Global state: Page number
 */
export const recoilPage = atom<number>({
  key: "pageAtom",
  default: 1,
});

/**
 * Global state: Max page number
 */
export const recoilMaxPage = atom<number>({
  key: "maxPageAtom ",
  default: 15,
});

/**
 * Global state: TTS enabled/disabled
 */
export const recoilTTS = atom({
  key: "ttsAtom",
  default: false, // TTS is disabled by default
});

/**
 * Function to safely parse JSON from storage
 * @param value  The value to be parsed
 * @param defaultValue  The default value to use if the value is not found
 * @returns  The parsed value or the default value
 */
function safeParse<T>(value: string | null, defaultValue: T): T {
  if (value === null) return defaultValue;
  try {
    const parsed = JSON.parse(value) as T;
    return parsed;
  } catch {
    return defaultValue;
  }
}

/**
 * Function to initialize state from storage
 * @param setStateFunction  The function to set the state
 * @param storage  socalstorage/sessionstorage
 * @param key   The key to use to retrieve the value from storage ex "filterBy"
 * @param defaultValue  The default value to use if the value is not found
 */
export function initializeStateFromStorage<T>(
  setStateFunction: (val: T) => void,
  storage: Storage,
  key: string,
  defaultValue: T,
) {
  const storedValue = storage.getItem(key);
  const parsedValue = safeParse(storedValue, defaultValue);
  setStateFunction(parsedValue);
}

/**
 *  Function to update storage on change
 * @param key  The key to use to retrieve the value from storage ex "filterBy"
 * @param value  The value to be stored
 * @param storage  socalstorage/sessionstorage
 */
export function updateStorageOnChange<T>(
  key: string,
  value: T,
  storage: Storage,
) {
  storage.setItem(key, JSON.stringify(value));
}
