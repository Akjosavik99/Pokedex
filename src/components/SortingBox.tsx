import Box from "@mui/material/Box";
import InputLabel from "@mui/material/InputLabel";
import MenuItem from "@mui/material/MenuItem";
import FormControl from "@mui/material/FormControl";
import Select from "@mui/material/Select";
import { SortBy } from "./Home";

type sortingBoxProps = {
  updateSort: React.Dispatch<React.SetStateAction<string>>;
  sortBy: string;
};

export default function SortingBox({ updateSort, sortBy }: sortingBoxProps) {
  return (
    <Box sx={{ minWidth: 120 }}>
      <FormControl fullWidth>
        <InputLabel id="demo-simple-select-label">Sort by</InputLabel>
        <Select
          labelId="demo-simple-select-label"
          id="demo-simple-select"
          value={sortBy}
          label="Age"
          onChange={(e) => updateSort(e.target.value)}
        >
          <MenuItem value={SortBy.A_Z}>A-Z</MenuItem>
          <MenuItem value={SortBy.Z_A}>Z-A</MenuItem>
          <MenuItem value={SortBy.BASE_EXPERIENCE_INCREASING}>
            Base experience increasing
          </MenuItem>
          <MenuItem value={SortBy.BASE_EXPERIENCE_DECREASING}>
            Base experience decreasing
          </MenuItem>
          <MenuItem value={SortBy.WEIGHT_INCREASING}>
            Weight increasing
          </MenuItem>
          <MenuItem value={SortBy.WEIGHT_DECREASING}>
            Weight decreasing
          </MenuItem>
          <MenuItem value={SortBy.NONE}>Random</MenuItem>
        </Select>
      </FormControl>
    </Box>
  );
}
