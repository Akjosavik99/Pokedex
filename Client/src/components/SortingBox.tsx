import * as React from "react";
import Button from "@mui/material/Button";
import Menu from "@mui/material/Menu";
import MenuItem from "@mui/material/MenuItem";
import Fade from "@mui/material/Fade";
import { Box } from "@mui/material";
import sortings from "../assets/Sortings";
import { recoilSortBy } from "../recoil/atoms";
import { useRecoilState } from "recoil";

export default function SortingBox() {
  const [sortBy] = useRecoilState(recoilSortBy);
  const [currentSort, setCurrentSort] = React.useState<string>(sortBy);
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  function updateSortBy(sort: string) {
    console.log(sort + " is clicked");
    sessionStorage.setItem("sortBy", JSON.stringify(sort));
    setCurrentSort(sort);
  }

  return (
    <Box>
      <h2>Sortings</h2>
      <Button
        id="fade-button"
        aria-controls={open ? "fade-menu" : undefined}
        aria-haspopup="true"
        aria-expanded={open ? "true" : undefined}
        onClick={handleClick}
      >
        Sorting
      </Button>
      <Menu
        id="fade-menu"
        MenuListProps={{
          "aria-labelledby": "fade-button",
        }}
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        TransitionComponent={Fade}
      >
        {sortings.map((sorting) => (
          <MenuItem
            key={sorting[1]}
            value={sorting[1]}
            onClick={() => {
              handleClose;
              updateSortBy(sorting[0]);
            }}
            style={{
              backgroundColor:
                currentSort === sorting[0] ? "lightblue" : "white",
            }}
          >
            {sorting[0]}
          </MenuItem>
        ))}

        <MenuItem
          onClick={() => {
            handleClose;
            updateSortBy("A-Z");
          }}
        >
          Reset
        </MenuItem>
      </Menu>
    </Box>
  );
}
