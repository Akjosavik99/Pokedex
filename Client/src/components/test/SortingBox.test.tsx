import { fireEvent, render } from "@testing-library/react";
import { test, expect, describe } from "vitest";
import SortingBox from "../sorting_and_filtering/SortingBox";
import { RecoilRoot } from "recoil";

describe("SortingBox", () => {
  test("Checks that all the sorting options are rendered", async () => {
    const setCurrentSorting = () => {};

    const { getByRole, getByText } = render(
      <RecoilRoot>
        <SortingBox setCurrentSorting={setCurrentSorting} />
      </RecoilRoot>,
    );

    const sortings = [
      "ID increasing (default)",
      "ID decreasing",
      "A-Z",
      "Z-A",
      "XP increasing",
      "XP decreasing",
      "kg increasing",
      "kg decreasing",
    ];
    fireEvent.click(getByRole("button", { name: "Choose Sorting" }));
    sortings.forEach((sorting) => {
      const option = getByText(sorting);
      expect(option).toBeTruthy();
    });
  });

  test("Checks if sorting is applied correctly", () => {
    let currentSorting = "";
    const setCurrentSorting = (newSort: React.SetStateAction<string>) => {
      if (typeof newSort === "function") {
        currentSorting = newSort(currentSorting);
      } else {
        currentSorting = newSort;
      }
    };

    const { getByRole, getAllByText } = render(
      <RecoilRoot>
        <SortingBox setCurrentSorting={setCurrentSorting} />
      </RecoilRoot>,
    );

    fireEvent.click(getByRole("button", { name: "Choose Sorting" }));
    const option = getAllByText("XP increasing");
    fireEvent.click(option[1]);
    expect(currentSorting).toBe("base_experience,1");

    const option2 = getAllByText("Z-A");
    fireEvent.click(option2[1]);
    expect(currentSorting).toBe("name,-1");

    const option3 = getAllByText("ID increasing (default)");
    fireEvent.click(option3[1]);
    expect(currentSorting).toBe("_id,1");
  });
});
