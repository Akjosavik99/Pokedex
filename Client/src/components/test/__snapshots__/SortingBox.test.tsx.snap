// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SortingBox > Snapshot test of sorting box 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    >
      <div
        class="MuiBox-root css-0"
      >
        <button
          aria-controls="fade-menu"
          aria-expanded="true"
          aria-haspopup="true"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-32nhe7-MuiButtonBase-root-MuiButton-root"
          data-testid="sort-list-button"
          id="fade-button"
          tabindex="0"
          type="button"
        >
          Choose Sorting
           
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-4uv323-MuiSvgIcon-root"
            data-testid="ArrowDropDownCircleOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m0-2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 13-4-4h8z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiMenu-root MuiPopover-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
      id="fade-menu"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiMenu-paper MuiPopover-paper MuiMenu-paper css-3dzjca-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
        tabindex="-1"
      >
        <ul
          aria-labelledby="fade-button"
          class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
          role="menu"
          tabindex="-1"
        >
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-focusVisible MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="ID increasing (default)"
            role="menuitem"
            style="color: black;"
            tabindex="0"
            value="_id,1"
          >
            ID increasing (default)
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="ID decreasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="_id,-1"
          >
            ID decreasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="A-Z"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="name,1"
          >
            A-Z
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="Z-A"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="name,-1"
          >
            Z-A
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="XP increasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="base_experience,1"
          >
            XP increasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="XP decreasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="base_experience,-1"
          >
            XP decreasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="kg increasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="weight,1"
          >
            kg increasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="kg decreasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="weight,-1"
          >
            kg decreasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
    <div
      aria-hidden="true"
    >
      <div
        class="MuiBox-root css-0"
      >
        <button
          aria-haspopup="true"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-32nhe7-MuiButtonBase-root-MuiButton-root"
          data-testid="sort-list-button"
          id="fade-button"
          tabindex="0"
          type="button"
        >
          Choose Sorting
           
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-4uv323-MuiSvgIcon-root"
            data-testid="ArrowDropDownCircleOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m0-2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 13-4-4h8z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiMenu-root MuiPopover-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
      id="fade-menu"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 0; webkit-transition: opacity 195ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 195ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="-1"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiMenu-paper MuiPopover-paper MuiMenu-paper css-3dzjca-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
        style="opacity: 0; webkit-transition: opacity 195ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 195ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
        tabindex="-1"
      >
        <ul
          aria-labelledby="fade-button"
          class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
          role="menu"
          tabindex="-1"
        >
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="ID increasing (default)"
            role="menuitem"
            style="color: black;"
            tabindex="0"
            value="_id,1"
          >
            ID increasing (default)
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="ID decreasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="_id,-1"
          >
            ID decreasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="A-Z"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="name,1"
          >
            A-Z
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="Z-A"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="name,-1"
          >
            Z-A
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="XP increasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="base_experience,1"
          >
            XP increasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="XP decreasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="base_experience,-1"
          >
            XP decreasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="kg increasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="weight,1"
          >
            kg increasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="kg decreasing"
            role="menuitem"
            style="color: black;"
            tabindex="-1"
            value="weight,-1"
          >
            kg decreasing
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="-1"
      />
    </div>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <button
          aria-haspopup="true"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-32nhe7-MuiButtonBase-root-MuiButton-root"
          data-testid="sort-list-button"
          id="fade-button"
          tabindex="0"
          type="button"
        >
          Choose Sorting
           
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-4uv323-MuiSvgIcon-root"
            data-testid="ArrowDropDownCircleOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m0-2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 13-4-4h8z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-0"
    >
      <button
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-32nhe7-MuiButtonBase-root-MuiButton-root"
        data-testid="sort-list-button"
        id="fade-button"
        tabindex="0"
        type="button"
      >
        Choose Sorting
         
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-4uv323-MuiSvgIcon-root"
          data-testid="ArrowDropDownCircleOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m0-2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 13-4-4h8z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
